### PR TITLE
Add the latest changes in Datascan API

### PR DESCRIPTION
**Pull request by evasharma**
Updating Datascan terraform support with the latest changes and additions in Datascan API.

If this PR is for Terraform, I acknowledge that I have:

- [X] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [X ] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [X ] [Generated Terraform providers](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/generate-providers.md), and ran [`make test` and `make lint`](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/run-provider-tests/#run-unit-tests) in the generated providers to ensure it passes unit and linter tests.
- [X] [Ran](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/develop/run-tests.md) relevant acceptance tests using my own Google Cloud project and credentials (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [X] Read [Write release notes](https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:new-resource
`google_dataplex_datascan`
```

### CODE REVIEW

File: mmv1/products/dataplex/Datascan.yaml [hunk 1]
```diff
@@ -413,6 +441,47 @@ properties:
...
+          The fields to include in data profile.
+          If not specified, all fields at the time of profile scan job execution are included, except for ones listed in `exclude_fields`.
+        properties:
+          - !ruby/object:Api::Type::Array
+            name: 'fieldNames'
+            description: |
+              Expected input is a list of fully qualified names of fields as in the schema.
+              Only top-level field names for nested fields are supported.
+              For instance, if 'x' is of nested field type, listing 'x' is supported but 'x.y.z' is not supported. Here 'y' and 'y.z' are nested fields of 'x'.
+            item_type: !ruby/object:Api::Type::String
```
Comment (ScottSuarez) [inline mmv1/products/dataplex/Datascan.yaml:471]:
```suggestion
            item_type: Api::Type::String
```

File: mmv1/products/dataplex/Datascan.yaml [hunk 2]
```diff
@@ -413,6 +441,47 @@ properties:
...
+          The fields to exclude from data profile.
+          If specified, the fields will be excluded from data profile, regardless of `include_fields` value.
+        properties:
+          - !ruby/object:Api::Type::Array
+            name: 'fieldNames'
+            description: |
+              Expected input is a list of fully qualified names of fields as in the schema.
+              Only top-level field names for nested fields are supported.
+              For instance, if 'x' is of nested field type, listing 'x' is supported but 'x.y.z' is not supported. Here 'y' and 'y.z' are nested fields of 'x'.
+            item_type: !ruby/object:Api::Type::String
```
Comment (ScottSuarez) [inline mmv1/products/dataplex/Datascan.yaml:484]:
```suggestion
            item_type: item_type: Api::Type::String
```

File: mmv1/products/dataplex/Datascan.yaml [hunk 3]
```diff
@@ -794,3 +899,26 @@ properties:
...
+                  Execution state for the BigQuery exporting.
+                values:
+                  - STATE_UNSPECIFIED
+                  - SUCCEEDED
+                  - FAILED
+                  - SKIPPED
+              - !ruby/object:Api::Type::String
+                name: 'message'
+                description: |
+                  Additional information about the BigQuery exporting.
```
Comment (ScottSuarez) [inline mmv1/products/dataplex/Datascan.yaml:924]: This whole block seems like it should be an [output](https://github.com/GoogleCloudPlatform/magic-modules/blob/50087bde9e5ae445fac5fa643c8f81c195ded68a/mmv1/api/type.rb#L39) field.

File: mmv1/products/dataplex/Datascan.yaml [hunk 4]
```diff
@@ -463,6 +555,19 @@ properties:
...
+                    A mutable name for the rule.
+                    The name must contain only letters (a-z, A-Z), numbers (0-9), or hyphens (-).
+                    The maximum length is 63 characters.
+                    Must start with a letter.
+                    Must end with a number or a letter.
+                - !ruby/object:Api::Type::String
+                  name: 'description'
+                  description: |
+                    Description of the rule.
+                    The maximum length is 1,024 characters.
```
Comment (ScottSuarez) [inline mmv1/products/dataplex/Datascan.yaml:570]: likely [output](https://github.com/GoogleCloudPlatform/magic-modules/blob/50087bde9e5ae445fac5fa643c8f81c195ded68a/mmv1/api/type.rb#L39) fields

File: mmv1/products/dataplex/Datascan.yaml [hunk 5]
```diff
@@ -424,6 +493,29 @@ properties:
...
+                  Execution state for the BigQuery exporting.
+                values:
+                  - STATE_UNSPECIFIED
+                  - SUCCEEDED
+                  - FAILED
+                  - SKIPPED
+              - !ruby/object:Api::Type::String
+                name: 'message'
+                description: |
+                  Additional information about the BigQuery exporting.
```
Comment (ScottSuarez) [inline mmv1/products/dataplex/Datascan.yaml:518]: [output](https://github.com/GoogleCloudPlatform/magic-modules/blob/50087bde9e5ae445fac5fa643c8f81c195ded68a/mmv1/api/type.rb#L39) only

File: mmv1/products/dataplex/Datascan.yaml [hunk 6]
```diff
@@ -413,6 +441,47 @@ properties:
...
+          - !ruby/object:Api::Type::NestedObject
+            name: 'bigQueryExport'
+            description: |
+              If set, results will be exported to the provided BigQuery table.
+            properties:
+              - !ruby/object:Api::Type::String
+                name: 'resultsTable'
+                description: |
+                  The BigQuery table to export DataProfileScan results to.
+                  Format:projects/{project}/datasets/{dataset}/tables/{table}
```
Comment (ScottSuarez) [inline mmv1/products/dataplex/Datascan.yaml:458]: Is it possible to test this field. It's okay if not just want to ensure we are conscious in it's exclusion.

**Review by ScottSuarez — CHANGES_REQUESTED** [review]

File: mmv1/products/dataplex/Datascan.yaml (hunk 3 above)
Comment (saurabh-net) [reply mmv1/products/dataplex/Datascan.yaml:924]: Note that we are looking to remove the DataProfileResult and DataQualityResult in this PR: https://github.com/GoogleCloudPlatform/magic-modules/pull/8583#issuecomment-1670332184

Based on how that discussion goes, we should look at only adding the export and rule name fields in the spec, and not in the result.

@evasharma

File: mmv1/products/dataplex/Datascan.yaml (hunk 1 above)
Comment (evasharma) [reply mmv1/products/dataplex/Datascan.yaml]: Thanks Fixed!

File: mmv1/products/dataplex/Datascan.yaml (hunk 2 above)
Comment (evasharma) [reply mmv1/products/dataplex/Datascan.yaml]: Thanks Fixed!

File: mmv1/products/dataplex/Datascan.yaml (hunk 3 above)
Comment (evasharma) [reply mmv1/products/dataplex/Datascan.yaml]: Thanks removed the fields added in this PR, the rest of the Result field to be removed in @saurabh-net PR.

File: mmv1/products/dataplex/Datascan.yaml (hunk 4 above)
Comment (evasharma) [reply mmv1/products/dataplex/Datascan.yaml]: Thanks removed the fields added in this PR, the rest of the Result field to be removed in @saurabh-net PR.

File: mmv1/products/dataplex/Datascan.yaml (hunk 6 above)
Comment (evasharma) [reply mmv1/products/dataplex/Datascan.yaml]: It is required field and has been tested.

File: mmv1/products/dataplex/Datascan.yaml [hunk 7]
```diff
@@ -409,10 +439,53 @@ properties:
...
+          - !ruby/object:Api::Type::NestedObject
+            name: 'bigqueryExport'
+            description: |
+              If set, results will be exported to the provided BigQuery table.
+            properties:
+              - !ruby/object:Api::Type::String
+                name: 'resultsTable'
+                description: |
+                  The BigQuery table to export DataProfileScan results to.
+                  Format://bigquery.googleapis.com/projects/PROJECT_ID/datasets/DATASET_ID/tables/TABLE_ID
```
Comment (ScottSuarez) [inline mmv1/products/dataplex/Datascan.yaml:462]: Are you able to supply a test for this field? If not that is okay I just want to ensure we are conscious in it's exclusion.

File: mmv1/products/dataplex/Datascan.yaml [hunk 8]
```diff
@@ -336,7 +366,7 @@ properties:
               allow_empty_object: true
               send_empty_value: true
               description: |
-                ColumnAggregate rule which evaluates whether the column has duplicates.
+                Row-level rule which evaluates whether each column value is unique.
```
Comment (ScottSuarez) [inline mmv1/products/dataplex/Datascan.yaml:369]:
```suggestion
                Row-level rule which evaluates whether each row value is unique.
```

suggestion (?) unsure if read correctly. I could be wrong here

File: mmv1/products/dataplex/Datascan.yaml [hunk 9]
```diff
@@ -250,10 +250,27 @@ properties:
...
+          - !ruby/object:Api::Type::NestedObject
+            name: 'bigqueryExport'
+            description: |
+              If set, results will be exported to the provided BigQuery table.
+            properties:
+              - !ruby/object:Api::Type::String
+                name: 'resultsTable'
+                description: |
+                  The BigQuery table to export DataQualityScan results to.
+                  Format://bigquery.googleapis.com/projects/PROJECT_ID/datasets/DATASET_ID/tables/TABLE_ID
```
Comment (ScottSuarez) [inline mmv1/products/dataplex/Datascan.yaml:273]: Are you able to supply a test for this field? If not that is okay I just want to ensure we are conscious in it's exclusion.

**Review by ScottSuarez — CHANGES_REQUESTED** [review]

File: mmv1/products/dataplex/Datascan.yaml (hunk 9 above)
Comment (evasharma) [reply mmv1/products/dataplex/Datascan.yaml:273]: This one requires an input table path for which the scan creator should have permission, would be tricky to set up. We can exclude it from our tests for now.

Comment (saurabh-net) [reply mmv1/products/dataplex/Datascan.yaml:273]: Does the Datascan Export API create datasets if not present either? If so, we can simply specify a BQ table path with the Project being interpolated, and the test-runner will be able to create it.

Something like  project/"<%= ctx[:test_env_vars]['project_name'] %>"/dataset/testDataScan/table/testTable
----

We could also create a BQ Table via Terraform itself and reference that as part of the test setup.

Comment (saurabh-net) [reply mmv1/products/dataplex/Datascan.yaml:273]: https://github.com/GoogleCloudPlatform/magic-modules/blob/051ffe1f2feed35899fbb8d0ac8c1c510f238ee9/mmv1/templates/terraform/examples/dataplex_asset_primary.tf.erb 

This is an example of a Dataplex Asset setting up other resources it needs before itself, and then referencing them.

Comment (saurabh-net) [reply mmv1/products/dataplex/Datascan.yaml:273]: https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/bigquery_table_iam#google_bigquery_table_iam_member. --> We can use this to grant the Dataplex P4SA write permissions to the Table.

**Comment by ScottSuarez** [comment]
If you can't get the bigquery export tests to work we can omit them

File: mmv1/products/dataplex/Datascan.yaml (hunk 9 above)
Comment (evasharma) [reply mmv1/products/dataplex/Datascan.yaml:273]: Discussed with @saurabh-net offline as well to add more test cases eventually for providing more coverage.

Comment (saurabh-net) [reply mmv1/products/dataplex/Datascan.yaml:273]: Looks like this proved harder than expected; agree on adding tests for this later.

**Review by ScottSuarez — APPROVED** [review]

**Pull request merged**